rust/src/ima.rs

```diff
@@ -0,0 +1,234 @@
+}
+
+impl CommitRewriter {
+    fn new(repo_path: &str) -> Result<Self> {
+        let repo = ostree::Repo::new_for_path(repo_path);
+        repo.open(gio::NONE_CANCELLABLE)?;
+        Ok(Self {
+            repo,
+            rewritten_files: Default::default(),
+        })
+    }
+
+    fn map_xattrs(&self, v: &glib::Variant) -> glib::Variant {
+        let v = v.get_data_as_bytes();
+        let v = v.try_as_aligned().unwrap();
+        let v = gv!("a(ayay)").cast(v);
+        let mut map: BTreeMap<Vec<u8>, Vec<u8>> = BTreeMap::new();
+        for e in v.iter() {
+            let (k, v) = e.to_tuple();
+            map.insert(k.into(), v.into());
+        }
+        let k = b"system.foo\0";
+        let v = b"bar\0";
+        map.insert((*k).into(), (*v).into());
+
+        let map: Vec<_> = map.into_iter().collect();
+        variant_utils::new_variant_a_ayay(&map)
+    }
+
+    fn map_file(&mut self, checksum: &str) -> Result<Rc<str>> {
+        if let Some(r) = self.rewritten_files.get(checksum) {
+            return Ok(Rc::clone(r));
+        }
+        let cancellable = gio::NONE_CANCELLABLE;
+        let (instream, meta, xattrs) = self.repo.load_file(checksum, cancellable)?;
+        let instream =
+            instream.unwrap_or_else(|| gio::MemoryInputStream::new().upcast::<gio::InputStream>());
+        let meta = meta.unwrap();
+        let xattrs = xattrs.unwrap();
+
+        let xattrs = self.map_xattrs(&xattrs);
+
+        let (ostream, size) =
+            ostree::raw_file_to_content_stream(&instream, &meta, Some(&xattrs), cancellable)?;
+        let new_checksum = self
+            .repo
+            .write_content(None, &ostream, size, cancellable)?
+            .to_hex();
+
+        let r: Rc<str> = new_checksum.into();
+        self.rewritten_files
+            .insert(checksum.to_string(), Rc::clone(&r));
+        Ok(r)
+    }
+
+    /// Write a dirtree object.
+    fn map_dirtree(&mut self, checksum: &str) -> Result<String> {
+        let src = &self
+            .repo
+            .load_variant(ostree::ObjectType::DirTree, checksum)?;
+        let src = src.get_data_as_bytes();
+        let src = src.try_as_aligned()?;
+        let src = gv!("(a(say)a(sayay))").cast(src);
+        let (files, dirs) = src.to_tuple();
+
+        // A reusable buffer to avoid heap allocating these
+        let mut hexbuf = [0u8; 64];
+
+        let new_files_builder =
+            unsafe { glib_sys::g_variant_builder_new(b"a(say)\0".as_ptr() as *const _) };
+        for file in files {
+            let (name, csum) = file.to_tuple();
+            let name = name.to_str();
+            hex::encode_to_slice(csum, &mut hexbuf)?;
+            let checksum = std::str::from_utf8(&hexbuf)?;
+            let mapped = self.map_file(checksum)?;
+            let mapped = hex::decode(&*mapped)?;
+            unsafe {
+                // Unwrap safety: The name won't have NULs
+                let name = CString::new(name).unwrap();
```

Comment:
Annoying to have to malloc the name again...I bet we could extend the `gvariant` crates to give us a `s` as as a C string pointer.

Reply:
I don't know much about how the file data are organized, but you'll have to calculate a hash (sha256) over the file data so that that hash can then be signed. I suppose this csum isn't 'it'.

Reply:
The ostree hash covers the file metadata (uid, gid, xattrs) *and* the content.  But that's not related to IMA.  AIUI `evmctl` is calculating that itself from the file we pass, right?

Reply:
yes, it will read all data from the filesystem:

https://sourceforge.net/p/linux-ima/ima-evm-utils/ci/master/tree/src/evmctl.c#l351

It will hash over extended attributes it finds (https://sourceforge.net/p/linux-ima/ima-evm-utils/ci/master/tree/src/evmctl.c#l79) and accumulate their values in a hash (https://sourceforge.net/p/linux-ima/ima-evm-utils/ci/master/tree/src/evmctl.c#l447) and then also consider uid, gid and mode from the stat (https://sourceforge.net/p/linux-ima/ima-evm-utils/ci/master/tree/src/evmctl.c#l456)

Reply:
Yep.  So the IMA and ostree hashes are very, very close in what they're hashing.  They might only differ in fact due to the ordering/structure of what's being hashed.